virtualmachine/aws/util.go

```diff
@@ -288,3 +288,105 @@ func DeleteKeyPair(name string, region string) error {
 
 	return nil
 }
+
+// GetInstanceStatus: returns status of given instances
+// Status includes availabilityZone & state
+func GetInstanceStatus(svc *ec2.EC2, instID string) (*InstanceStatus, error) {
+	input := &ec2.DescribeInstanceStatusInput{
+		IncludeAllInstances: func(val bool) *bool { return &val }(true),
+		InstanceIds:         []*string{&instID}}
+	statusOutput, err := svc.DescribeInstanceStatus(input)
+
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get instance status: %v", err)
+	}
+
+	instanceStatus := statusOutput.InstanceStatuses[0]
+	status := &InstanceStatus{
+		AvailabilityZone: *instanceStatus.AvailabilityZone,
+		InstanceId:       *instanceStatus.InstanceId,
+		State:            *instanceStatus.InstanceState.Name}
+
+	return status, nil
+}
+
+// GetFilters: converts filter map to array of pointers to ec2.Filter objects
+func GetFilters(filterMap map[string][]*string) []*ec2.Filter {
+	filters := make([]*ec2.Filter, 0)
+	for key, values := range filterMap {
+		filters = append(filters, &ec2.Filter{
+			Name:   &key,
+			Values: values})
+	}
+	return filters
+}
+
+// ToEc2IpPermissions: converts array of IpPermission objects to
+// array of pointer to ec2.IpPermission
+func ToEc2IpPermissions(ipPermissions []IpPermission) []*ec2.IpPermission {
+	ec2IpPermissions := make([]*ec2.IpPermission, 0)
+	for _, ipPermission := range ipPermissions {
+		ec2IpPermission := &ec2.IpPermission{
+			FromPort:   ipPermission.FromPort,
+			ToPort:     ipPermission.ToPort,
+			IpProtocol: &ipPermission.IpProtocol}
+
+		if ipPermission.Ipv4Ranges != nil && len(ipPermission.Ipv4Ranges) > 0 {
+			ipRanges := make([]*ec2.IpRange, 0)
+			for _, ipv4Range := range ipPermission.Ipv4Ranges {
+				ipRanges = append(ipRanges, &ec2.IpRange{
+					CidrIp: &ipv4Range})
+			}
+			ec2IpPermission.IpRanges = ipRanges
+		} else {
```

Comment:
do we need else here? can we have both permissions?

Reply:
Yes, because we can not specify permission for both , ipv4Blocks and ipv6Blocks, in same block. AWS sdk with throw error in that case. Should this be mentioned in comment?